### PR TITLE
Add array constraint generation

### PR DESCRIPTION
**Pull request by iangneal**
Generates constraints for each array element when `fr_copy_n` is used. For an array of N elements, this will create N calls to `__constraint_values`.

### CODE REVIEW

File: compiler/src/intermediate_representation/constraint_bucket.rs [hunk 1]
```diff
@@ -112,21 +114,72 @@ impl WriteLLVMIR for ConstraintBucket {
...
-                    &[
-                        to_basic_metadata_enum(lhs),
-                        to_basic_metadata_enum(rhs),
-                        to_basic_metadata_enum(constr),
-                    ],
-                );
-                Some(call)
+            ConstraintBucket::Substitution(i) => {
+                let size = match i.as_ref() {
+                    Instruction::Value(_) => todo!(),
```
Comment (kferles) [inline compiler/src/intermediate_representation/constraint_bucket.rs:119]: Not familiar enough to review this part of the coebase, but what are all these TODOs? If they are current limitations, please document them.

File: compiler/src/intermediate_representation/constraint_bucket.rs [hunk 2]
```diff
@@ -112,21 +114,61 @@ impl WriteLLVMIR for ConstraintBucket {
...
+                            CONSTRAINT_VALUES_FN_NAME,
+                            &[
+                                to_basic_metadata_enum(lhs),
+                                to_basic_metadata_enum(rhs),
+                                to_basic_metadata_enum(constr),
+                            ],
+                        )
+                    }).collect();
+
+                    Some(constraint_calls.last().expect("must have >1 value!").clone())
```
Comment (tim-hoffman) [inline compiler/src/intermediate_representation/constraint_bucket.rs:169]: I don't think we want a clone here. If that causes a borrow check issue, then you probably need to convert the map above to a normal loop like the BlockBucket implementation.

**Review by tim-hoffman — CHANGES_REQUESTED** [review]

File: compiler/src/intermediate_representation/constraint_bucket.rs (hunk 2 above)
Comment (iangneal) [reply compiler/src/intermediate_representation/constraint_bucket.rs]: Gotcha. I got overzealous with fun rust features.

**Review by tim-hoffman — APPROVED** [review]
LGTM

**Pull request merged**